File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        stack = []
        curr=head
        ret = None
        while curr:
            stack.append(curr)
            curr=curr.next
        
        node_to_remove = None

        for i in range(0,n):
            node_to_remove = stack.pop()
        
        print(node_to_remove.val)
        if stack:
            node_before = stack.pop()

            node_before.next = node_to_remove.next
            node_to_remove.next = None
            ret = head
        else:
            ret = head.next
    
        return ret




        

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* dummy = new ListNode();
        ListNode* curr = dummy;

        ListNode* p1 = list1;
        ListNode* p2 = list2;

        while (p1 != nullptr and p2 != nullptr) {
            if (p1->val <= p2->val) {
                ListNode* saved = p1->next;
                p1->next = nullptr;
                curr->next = p1;
                curr = curr->next;
                p1 = saved;
            } else {
                ListNode* saved = p2->next;
                p2->next = nullptr;
                curr->next = p2;
                curr = curr->next;
                p2 = saved;
            }
        }
        if (p2 != nullptr) {
            curr->next = p2;
        }

        if (p1 != nullptr) {
            curr->next = p1;
        }

        return dummy->next;


    }
};
